src/Components/Web.Extensions/src/wwwroot/inputFile.js

```diff
@@ -0,0 +1,142 @@
+            // Permits replacing an existing file with a new one of the same file name.
+            elem.value = '';
+        });
+
+        elem.addEventListener('change', function () {
+            // Reduce to purely serializable data, plus an index by ID.
+            elem._blazorFilesById = {};
+
+            const fileList = Array.prototype.map.call(elem.files, function (file) {
+                const result = {
+                    id: ++elem._blazorInputFileNextFileId,
+                    lastModified: new Date(file.lastModified).toISOString(),
+                    name: file.name,
+                    size: file.size,
+                    type: file.type,
+                };
+
+                elem._blazorFilesById[result.id] = result;
+
+                // Attach the blob data itself as a non-enumerable property so it doesn't appear in the JSON.
+                Object.defineProperty(result, 'blob', { value: file });
+
+                return result;
+            });
+
+            callbackWrapper.invokeMethodAsync('NotifyChange', fileList);
+        });
+    }
+
+    function toImageFile(elem, fileId, format, maxWidth, maxHeight) {
+        var originalFile = getFileById(elem, fileId);
+
+        return new Promise(function (resolve) {
+            var originalFileImage = new Image();
+            originalFileImage.onload = function () { resolve(originalFileImage); };
+            originalFileImage.src = URL.createObjectURL(originalFile.blob);
+        }).then(function (loadedImage) {
+            return new Promise(function (resolve) {
+                var desiredWidthRatio = Math.min(1, maxWidth / loadedImage.width);
+                var desiredHeightRatio = Math.min(1, maxHeight / loadedImage.height);
+                var chosenSizeRatio = Math.min(desiredWidthRatio, desiredHeightRatio);
+
+                var canvas = document.createElement('canvas');
+                canvas.width = Math.round(loadedImage.width * chosenSizeRatio);
+                canvas.height = Math.round(loadedImage.height * chosenSizeRatio);
+                canvas.getContext('2d').drawImage(loadedImage, 0, 0, canvas.width, canvas.height);
+                canvas.toBlob(resolve, format);
+            });
+        }).then(function (resizedImageBlob) {
+            var result = {
+                id: ++elem._blazorInputFileNextFileId,
+                lastModified: originalFile.lastModified,
+                name: originalFile.name, // Note: we're not changing the file extension.
+                size: resizedImageBlob.size,
+                type: format,
+                relativePath: originalFile.relativePath
+            };
+
+            elem._blazorFilesById[result.id] = result;
+
+            // Attach the blob data itself as a non-enumerable property so it doesn't appear in the JSON.
+            Object.defineProperty(result, 'blob', { value: resizedImageBlob });
+
+            return result;
+        });
+    }
```

Comment:
With TransformAsync this becomes:
```
function transformFile(transformer){
        var originalFile = getFileById(elem, fileId);
        return new Promise(function (resolve) {
              return transformer(originalFile.blob).then(newBlob => resolve(newBlob));     
        }).then(function (newBlob) {
            var result = {
                id: ++elem._blazorInputFileNextFileId,
                lastModified: originalFile.lastModified,
                name: originalFile.name, // Note: we're not changing the file extension.
                size: resizedImageBlob.size,
                type: format,
                relativePath: originalFile.relativePath
            };

            elem._blazorFilesById[result.id] = result;

            // Attach the blob data itself as a non-enumerable property so it doesn't appear in the JSON.
            Object.defineProperty(result, 'blob', { value: newBlob});

            return result;
        });
    }
}
```

And it can largely be simplified

Reply:
Then we can tell customers who need it to implement the image manipulation as a transformer